# <center> Production Optimization Example
## <center> Systems Engineering and Analysis
## <center> <img src="https://www.engr.colostate.edu/~jdaily/Systems-EN-CSU-1-C357.svg" width="400" /> 
### <center> Prepared by: Dr. Jeremy Daily

### Constrained Optimization with Linear Constraints
This is the example from Section 9.5 in B&F.
See Table 9.11 for the input data for the problem.

In [28]:
import numpy as np
from scipy import optimize

We'll use the minimize function from the optimize module of the scipy library.

https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html

In [39]:
# Define a cost function. In this case we want to maximize profit,
# so we have to make the cost-function negative to convert it to a 
# minimization problem supported by scipy.optimize.minimize
def cost_function(x):
    """
    Pass in a vector of length 2 representing the design variables.
    """
    nA = x[0]
    nB = x[1]
    #Make sure it is negative so we can minimize the function to 
    # determine maximum profits
    return -(nA*0.60 + nB*0.70)

In [40]:
#Initial guess as to the quanity of each A and B.
x0 = np.array([1,1])

In [41]:
# Specify the bounds on the design variables.
# In this case, we don't want to produce negative quantities.
bounds = optimize.Bounds([0,0],[np.inf, np.inf])

Linear constraints have a matrix form:
$$\mathbf{A}\vec{x}\le\vec{c}$$
In this case,
$$A = \begin{bmatrix}
2.4 & 3 \\
0 & 2.5 \\
5 & 0 \\
\end{bmatrix}$$
and 
$$ \vec{c} = \begin{Bmatrix} 1200 \\600\\1500 \end{Bmatrix}$$

In [42]:
# Build a constraints array
A = np.array([[2.4,3],[0,2.5],[5,0]])
# Assign lower and upper bounds to the Ax product.
constraints = optimize.LinearConstraint(A,[0,0,0],[1200,600,1500])
print(A)

[[2.4 3. ]
 [0.  2.5]
 [5.  0. ]]


In [43]:
# Check to see the constraint calculation
A.dot(x0)

array([5.4, 2.5, 5. ])

In [44]:
# Calculate a solution
sol = optimize.minimize(cost_function, x0, bounds=bounds, constraints=constraints)
sol

     fun: -291.99999999968236
     jac: array([-0.59999847, -0.70000076])
 message: 'Optimization terminated successfully'
    nfev: 39
     nit: 13
    njev: 13
  status: 0
 success: True
       x: array([300., 160.])

Note the solution is nA = 300 and nB = 160. The algorithm is satisfied that it found a minimum. Try changing the objective function by removing the - and try again. What happens?

In [45]:
# Compute the constraints
A.dot(sol.x)

array([1200.,  400., 1500.])

In [47]:
print("For maximum profit, produce {:.0f} of A and {:.0f} of B.".format(sol.x[0],sol.x[1]))

For maximum profit, produce 300 of A and 160 of B.
